src/main.cpp

```diff
@@ -4619,54 +4600,72 @@ bool SendMessages(CNode* pto, bool fSendTrickle)
         //
         // Message: inventory
         //
-        vector<CInv> vInv;
-        vector<CInv> vInvWait;
-        {
-            LOCK(pto->cs_inventory);
-            vInv.reserve(pto->vInventoryToSend.size());
-            vInvWait.reserve(pto->vInventoryToSend.size());
-            BOOST_FOREACH(const CInv& inv, pto->vInventoryToSend)
-            {
-                if (pto->setInventoryKnown.count(inv))
-                    continue;
+        vector<CInv> vInv, vInvWait;
+        nNowMicros = GetTimeMicros();
 
-                // trickle out tx inv to protect privacy
-                if (inv.type == MSG_TX && !fSendTrickle)
-                {
-                    // 1/4 of tx invs blast to all immediately
-                    static uint256 hashSalt;
-                    if (hashSalt.IsNull())
-                        hashSalt = GetRandHash();
-                    uint256 hashRand = ArithToUint256(UintToArith256(inv.hash) ^ UintToArith256(hashSalt));
-                    hashRand = Hash(BEGIN(hashRand), END(hashRand));
-                    bool fTrickleWait = ((UintToArith256(hashRand) & 3) != 0);
+        vInv.reserve(1000);
+        
+        if (pto->nNextInvSend < nNowMicros)
+            vInvWait.reserve(pto->vInventoryToSend.size());
 
-                    if (fTrickleWait)
+        {
+            LOCK(pto->cs_inventory);
+            
+            // TODO it would be sufficient to walk vInventoryToSend randomly, shuffling is relatively inefficient
+            std::random_shuffle(pto->vInventoryToSend.begin(), pto->vInventoryToSend.end());
+
+            BOOST_FOREACH(const CInv& inv, pto->vInventoryToSend) {
+                switch(inv.type) {
+                    case MSG_BLOCK:
+                    case MSG_FILTERED_BLOCK:
                     {
-                        vInvWait.push_back(inv);
-                        continue;
+                        // Aggressively push MSG_BLOCK/MSG_FILTERED_BLOCK
```

Comment:
Maybe it is more efficient to keep a separate vInventoryToSend for MSG_BLOCK and MSG_FILTERED_BLOCK, and only shuffle/process the normal one when nNextInvSend is exceeded.